apps/gdalalg_raster_calc.cpp

```diff
@@ -0,0 +1,619 @@
+                                           inBandVariable.c_str());
+
+                CPLXMLNode *sourceFilename =
+                    CPLCreateXMLNode(source, CXT_Element, "SourceFilename");
+                CPLAddXMLAttributeAndValue(sourceFilename, "relativeToVRT",
+                                           "0");
+                CPLCreateXMLNode(sourceFilename, CXT_Text, dsn.c_str());
+
+                CPLXMLNode *sourceBand =
+                    CPLCreateXMLNode(source, CXT_Element, "SourceBand");
+                CPLCreateXMLNode(sourceBand, CXT_Text,
+                                 std::to_string(nInBand).c_str());
+
+                // TODO add <SourceProperties> ?
+
+                CPLXMLNode *srcRect =
+                    CPLCreateXMLNode(source, CXT_Element, "SrcRect");
+                CPLAddXMLAttributeAndValue(srcRect, "xOff", "0");
+                CPLAddXMLAttributeAndValue(srcRect, "yOff", "0");
+                CPLAddXMLAttributeAndValue(srcRect, "xSize",
+                                           std::to_string(props.nX).c_str());
+                CPLAddXMLAttributeAndValue(srcRect, "ySize",
+                                           std::to_string(props.nY).c_str());
+
+                CPLXMLNode *dstRect =
+                    CPLCreateXMLNode(source, CXT_Element, "DstRect");
+                CPLAddXMLAttributeAndValue(dstRect, "xOff", "0");
+                CPLAddXMLAttributeAndValue(dstRect, "yOff", "0");
+                CPLAddXMLAttributeAndValue(dstRect, "xSize",
+                                           std::to_string(nXOut).c_str());
+                CPLAddXMLAttributeAndValue(dstRect, "ySize",
+                                           std::to_string(nYOut).c_str());
+            }
+        }
+
+        // Add the expression as a last step, because we may modify the
+        // expression as we iterate through the bands.
+        CPLAddXMLAttributeAndValue(arguments, "expression",
+                                   bandExpression.c_str());
+        CPLAddXMLAttributeAndValue(arguments, "dialect", "muparser");
+    }
+
+    return true;
+}
+
+static bool ParseSourceDescriptors(const std::vector<std::string> &inputs,
+                                   std::map<std::string, std::string> &datasets,
+                                   std::string &firstSourceName)
+{
+    bool isFirst = true;
+
+    for (const auto &input : inputs)
+    {
+        std::string name = "";
+
+        auto pos = input.find('=');
+        if (pos == std::string::npos)
+        {
+            if (inputs.size() > 1)
+            {
+                CPLError(CE_Failure, CPLE_AppDefined,
+                         "Inputs must be named when more than one input is "
+                         "provided.");
+                return false;
+            }
+            name = "X";
+        }
+        else
+        {
+            name = input.substr(0, pos);
+        }
+
+        std::string dsn = input.substr(pos + 1, input.size());
```

Comment:
This causes an issue with unsigned-integer-overflow sanitizer when pos == std::string::npos:
```
 /home/runner/work/gdal/gdal/apps/gdalalg_raster_calc.cpp:371:44: runtime error: unsigned integer overflow: 18446744073709551615 + 1 cannot be represented in type 'size_type' (aka 'unsigned long')
```
and the use of input.size() is probably OK, but not needed, so I guess this should be:

```suggestion
        std::string dsn = (pos == std::string::npos) ? input : input.substr(pos + 1);
```